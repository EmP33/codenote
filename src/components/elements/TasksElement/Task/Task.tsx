import React, { useState, useEffect } from "react";
import { Box, Checkbox, Typography } from "@mui/material";
import { useAppDispatch, useAppSelector } from "../../../../lib/hooks";
import { changeTaskStatus } from "../../../../store/user-slice";

interface TaskProps {
  title: string;
  date: string;
  status: string;
  id: string;
}

const Task: React.FC<TaskProps> = ({ title, date, status, id }) => {
  const dispatch = useAppDispatch();
  const user = useAppSelector((state) => state.user.user);
  const [checked, setChecked] = useState(status === "completed");
  const dueDate = new Date(date);

  useEffect(() => {
    setChecked(status === "completed");
  }, [status]);

  const changeStatusHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    let status: string;
    if (e.target.checked) {
      // If is checked
      status = "completed";
    } else {
      // If isnt
      status = "progress";
    }
    dispatch(
      changeTaskStatus(user.uid, {
        title: title,
        status: status,
        date: date,
        id: id,
      })
    );
  };
  return (
    <Box
      sx={{
        background: "#fff",
        color: "var(--color-primary)",
        display: "grid",
        gridTemplateColumns: "repeat(2,minmax(min-content,max-content)) 1fr",
        alignItems: "center",
        padding: 0.5,
        marginBottom: 3,
        borderRadius: 1,
      }}
    >
      <Checkbox
        checked={checked}
        color="secondary"
        onChange={changeStatusHandler}
      />
      <Typography
        sx={{
          fontSize: { xs: 12, sm: 14, md: 16 },
          textDecoration: checked ? "line-through" : "none",
        }}
      >
        {title}
      </Typography>
      <Typography
        sx={{
          justifySelf: "flex-end",
          marginRight: 1,
          fontSize: { xs: 12, sm: 14, md: 16 },
        }}
        color="secondary"
      >
        {dueDate.getFullYear()}/{dueDate.getMonth() + 1 <= 9 && "0"}
        {dueDate.getMonth() + 1}/{dueDate.getDate() <= 9 && "0"}
        {dueDate.getDate()}
      </Typography>
    </Box>
  );
};

export default Task;
