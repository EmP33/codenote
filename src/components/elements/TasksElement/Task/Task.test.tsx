import { shallow } from "enzyme";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Task from "./Task";

describe("Task testing", () => {
  it("should render Task component properly", () => {
    expect(
      shallow(
        <Task
          id="t1"
          status="test"
          title="Example Title"
          date="Due Today, 2:00 PM"
          pinnedNote={{ blocks: [], id: "n1", date: 123, views: 1 }}
        />
      )
    ).toMatchSnapshot();
  });
  it("should render proper title", () => {
    render(
      <Task
        id="t2"
        status="test"
        title="Example Title"
        date="Due Today, 2:00 PM"
        pinnedNote={{ blocks: [], id: "n2", date: 123, views: 1 }}
      />
    );
    const task = screen.getByText("example title", { exact: false });
    expect(task).toBeInTheDocument();
  });
  it("should render proper date", () => {
    render(
      <Task
        id="t3"
        status="test"
        title="Example Title"
        date="Due Today, 2:00 PM"
        pinnedNote={{ blocks: [], id: "n3", date: 123, views: 1 }}
      />
    );
    const task = screen.getByText("Due Today, 2:00 PM", { exact: false });
    expect(task).toBeInTheDocument();
  });
});
