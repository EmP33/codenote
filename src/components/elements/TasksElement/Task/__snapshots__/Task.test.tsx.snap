// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task testing should render Task component properly 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "alignItems": "center",
      "background": "#fff",
      "borderRadius": 1,
      "color": "var(--color-primary)",
      "display": "grid",
      "gridTemplateColumns": "repeat(2,minmax(min-content,max-content)) 1fr",
      "marginBottom": 3,
      "padding": 0.5,
    }
  }
>
  <ForwardRef(Checkbox)
    color="secondary"
  />
  <ForwardRef(Typography)
    sx={
      Object {
        "fontSize": Object {
          "md": 16,
          "sm": 14,
          "xs": 12,
        },
      }
    }
  >
    Example Title
  </ForwardRef(Typography)>
  <ForwardRef(Typography)
    color="secondary"
    sx={
      Object {
        "fontSize": Object {
          "md": 16,
          "sm": 14,
          "xs": 12,
        },
        "justifySelf": "flex-end",
        "marginRight": 1,
      }
    }
  >
    Due Today, 2:00 PM
  </ForwardRef(Typography)>
</ForwardRef(Box)>
`;
